CZI EOSS 5 Notebook v7 maintenance and documentation (#140)
jupyterlab/team-compass: grants/czi-eoss5-notebook-v7-loi.md

## Proposal Title: Auto-filled; Maximum of 60 characters, including spaces.

Jupyter Notebook maintenance and community support for v4 migration.

## Amount Requested: Total budget amount requested in USD, including indirect costs; this number should be between $100,000 USD and $400,000 USD total costs over a two-year period.

$400,000 USD

## Proposal Summary/Scope of Work: Provide a short summary of the work being proposed (maximum of 500 words).

> JupyterLab and the Classic Notebook are the two major user interfaces to manipulate the Jupyter notebook documents. That format is really popular within the scientific community thanks to its ability to bring code, data analysis and visualization in a single document.

Reviewer: ```suggestion
JupyterLab and the Classic Notebook are the two major user interfaces to manipulate Jupyter notebook documents. This document format is popular within the scientific community because it brings code, data analysis, and visualization together in one document.
```

> This proposal follows the major the upcoming major version of the Classical Notebook will be built on top of JupyterLab plugins (that use newer web technologies). The three key ideas approved in the related migration enhancement proposal are: document-centric user experience, extensions are critical to the community and existing educational content relies on the current Notebook experience. The user interface changes try to be limited (although new features will require adaptation). But the extension API will be totally different.

Reviewer: ```suggestion
This proposal is for the upcoming major version of the Classic Notebook that will be built with JupyterLab plugins that use newer web technologies. The three key ideas approved in the related migration enhancement proposal are:

1. Document-centric user experience
2. Extensions are critical to the community
3. Existing educational content relies on the current Notebook experience

We will try to minimize the user interface changes although new features will require adaptation. The extension API will be totally different.
```

> As maintenance is mainly done by a couple of volunteers during their free time. The backlog of issues is ever growing due to the lack of manpower. Addressing that, supporting extensions migration and the new user issues for the new version will require an important effort.

Reviewer: ```suggestion
The backlog of issues is growing because maintenance is done by only a couple of volunteers during their spare time. It is important to address this backlog by supporting extensions migration and by anticipating new user issues for the new software versions.
```

To mitigate those points, the actions in the grant frame can be grouped into two categories: documentation and code base. Here are the lists of envisaged actions in priority order.

- Documentation
  - Generate screencasts automatically.
  - Today the screencasts are recorded manually and published on YouTube. This complicates their updates, their translation and their availability for educators.

>   - To produce screencasts we will create a set of helpers to emulate keyboard and mouse interactions (to be published as a stand-alone package to allow its usage in other projects). Then the video scenarii will need to be defined. And finally an automatic workflow to generate the videos will be set up.

Reviewer: ```suggestion
  - To produce screencasts, we will create a set of helpers to emulate keyboard and mouse interactions (to be published as a stand-alone package to allow its usage in other projects). Then, we will need to define the video scenarios. Finally, we will set up an automatic workflow to generate the videos.
```

>   - Refactor the existing documentation focusing on

Reviewer: ```suggestion
  - Refactor the existing documentation, focusing on:
```

>     - Reduce fragmentation of the documentation spread it multiple websites

Reviewer: ```suggestion
    - Centralize documentation that is currently spread across multiple websites
```

- Code base
  - Support the community extensions authors to migrate their code
  - Address issues related to backward inconsistency in the new major classical notebook version
  - Reduce the issues backlog focusing on bug fixes and features requested by the biomedical community based on the 2021 survey; at the time of this proposal writing, the document generation features and easier extension creation are the candidates.

## Value to Biomedical Users: Describe the expected value of the proposed work to the biomedical research community (maximum of 250 words).

JupyterLab and Notebook are primary tools used for biomedical science education, research, and manufacturing. Their interface provides a set of tools allowing users to work with code and data in a common document. Jupyter’s growth in schools, universities and laboratories speak to the value of these technologies in science.
Moving to newer technologies for the classical notebook is required to ensure the maintenance cost stays manageable and the tool will keep up with new browser features and requirements.

> Keeping improving the tool focusing on research code documentation will improve spreading research results and understanding of the new discoveries.

Reviewer: ```suggestion
Keep improving the tool, focusing on research code documentation, to facilitate sharing research results and to foster an understanding of new discoveries.
```
